src/dune_rules/pkg_rules.ml

```diff
@@ -193,7 +195,114 @@ module Value_list_env = struct
   let add_path (t : t) var path : t =
     Env.Map.update t var ~f:(fun paths ->
       let paths = Option.value paths ~default:[] in
-      Some (Value.Dir (Path.build path) :: paths))
+      Some (Value.Dir path :: paths))
+  ;;
+
+  (* Prepends the bin dir of a given toolchain compiler to the PATH
+     variable in the environment. *)
+  let add_toolchain_bin_dir_to_path t compiler =
+    let toolchain_bin_dir =
+      Toolchain.Compiler.bin_dir compiler |> Path.outside_build_dir
+    in
+    add_path t Env_path.var toolchain_bin_dir
+  ;;
+end
+
+module DB = struct
+  type t =
+    { all : Lock_dir.Pkg.t Package.Name.Map.t
+    ; system_provided : Package.Name.Set.t
+    }
+
+  let equal t { all; system_provided } =
+    Package.Name.Map.equal ~equal:Lock_dir.Pkg.equal t.all all
+    && Package.Name.Set.equal t.system_provided system_provided
+  ;;
+
+  let get =
+    let dune = Package.Name.Set.singleton (Package.Name.of_string "dune") in
+    fun context ->
+      let+ all = Lock_dir.get context in
+      { all = all.packages; system_provided = dune }
+  ;;
+end
+
+module Compiler = struct
```

Comment:
Can this module be extracted to its own file?

Reply:
The value would be that it would make it easier to localize the impact of this feature to one module. Following that, a written interface for it that would give us a complete picture of how this feature interacts with the rules.

Looking at the code, it needs a bit of improvement to extract out such an interface. For example, this excerpt is an implementation detail of this feature and the rules have no business knowing anything about it:

```ocaml
      Memo.of_reproducible_fiber
        (let open Fiber.O in
         let* available_compilers =
           Toolchain.Available_compilers.load_upstream_opam_repo ()
         in
         let* toolchain_compiler =
           Toolchain.Available_compilers.find available_compilers name version ~deps
         in
         match toolchain_compiler with
         | None -> Fiber.return None
         | Some toolchain_compiler ->
           let+ toolchain_compiler =
             Toolchain.Compiler.ensure_installed
               ~log_when:`Install_only
               toolchain_compiler
           in
           Some toolchain_compiler)
```

I can make the changes myself if you aren't sure what I'm trying to get at.

Reply:
I did this change myself. A separate module wasn't necessary